TRUNK-6329: Modernize OpenMRS installation wizard UI to match the current OpenMRS branding

mherman22 opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111: Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
The current OpenMRS installation wizard could benefit from a more modern and intuitive UI that aligns better with OpenMRS branding. While we don't necessarily need to move away from using Velocity templates, the goal is to enhance the visual appeal and make the installation process more explicit and user-friendly.

## Issue I worked on
This pull request fixes the following tickets:
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first! -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it! -->
<!--- Just add the issue number at the end: -->
- https://issues.openmrs.org/browse/TRUNK-6329
- https://openmrs.atlassian.net/browse/TRUNK-6330
- https://openmrs.atlassian.net/browse/TRUNK-6333

## Summary of changes
- Improved the UI/UX appearance/feel of the installation wizard pages using improved HTML structure and modern styles but maintaining the apache velocity templates technology.
- Fix Random Mysql 'using password: NO' setup errors caused by the browser.

## Short videos

### This demonstrates the simplesetup installation mode

https://github.com/user-attachments/assets/33e804b0-6831-4386-a791-4a3fdcbff45b

### This demonstrates what happens incase of a startup error

https://github.com/user-attachments/assets/719aec58-9fbb-42a9-a4f2-2458bf85b86b

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

ibacher commented:
Nice work!

ibacher commented on web/src/main/java/org/openmrs/web/filter/startuperror/StartupErrorFilterModel.java:
```diff
@@ -24,6 +24,12 @@ public class StartupErrorFilterModel {
+	public static final String HEADER_TEMPLATE = "org/openmrs/web/filter/startuperror/header.vm";
+	
+	// automatically given to the .vm files and used there
+	public static final String FOOTER_TEMPLATE = "org/openmrs/web/filter/initialization/footer.vm";
```
So we have a common footer for everything? Wouldn't it make sense to move that template up a folder so it isn't part of the rest of the `initialization` files?

ibacher commented on web/src/main/java/org/openmrs/web/filter/util/SessionModelUtils.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public class SessionModelUtils {
+	
+	private static final org.slf4j.Logger log = LoggerFactory.getLogger(SessionModelUtils.class);
```
Is there a reason we don't import this?

ibacher commented on web/src/main/java/org/openmrs/web/filter/util/SessionModelUtils.java:
```diff
@@ -0,0 +1,102 @@
+	 * @param model the {@link InitializationWizardModel} to save
+	 */
+	public static void saveToSession(HttpSession session, InitializationWizardModel model) {
+		if (session == null || model == null) return;
```
Per our coding conventions, even one-line ifs should have curly braces.

```suggestion
		if (session == null || model == null) {
			return;
		}
```

ibacher commented on web/src/main/resources/org/openmrs/web/filter/initialization/installmethod.vm:
```diff
@@ -1,64 +1,257 @@
+</div>
+
+<style>
+	    body {
```
```suggestion
    body {
```

ibacher commented on web/src/main/resources/org/openmrs/web/filter/initialization/installmethod.vm:
```diff
@@ -1,64 +1,257 @@
+        background: white;
+        border-radius: 12px;
+        box-shadow: 0 2px 4px rgba(0,0,0,0.05);
+		height: calc(100vh - 180px);
```
```suggestion
        height: calc(100vh - 180px);
```

ibacher commented on webapp/src/main/webapp/index.htm:
```diff
@@ -1,51 +1,194 @@
-            height:45px;
-            width:100%;
-            font-size:24px;
+	<style>
```
```suggestion
        <style>
```

ibacher commented on webapp/src/main/webapp/index.htm:
```diff
@@ -1,51 +1,194 @@
-            width:100%;
-            font-size:24px;
+	<style>
+        :root {
```
I think the rules here should be intended four more spaces?

ibacher reviewed (commented):
Basically looks good. A few minor things to cleanup. Try to ensure we're consistently using either tabs or space to indent the Velocity templates. I don't think it matters too much, but it will make things cleaner.

mherman22 commented on web/src/main/java/org/openmrs/web/filter/startuperror/StartupErrorFilter.java:
```diff
@@ -32,20 +32,22 @@
 	 */
 	private static final String DEFAULT_PAGE = "generalerror.vm";
+
+	public static final String OPENMRS_VERSION = OpenmrsConstants.OPENMRS_VERSION_SHORT;
```
Oversight, i intended to remove it

mherman22 commented:
> @mherman22 did you also notice that the revision number is not replaced when running with jetty? `The OpenMRS 2.8.0-${revisionNumber} server is currently in maintenance mode.`

Yes I have, I should have  a followup commit to fix any pending issues including that.

dkayiwa commented:
@mherman22 with these changes, i have again got this error during setup: https://pastebin.com/kDMJu07Q
Do you wanna fix it too?

mherman22 commented:
> @mherman22 with these changes, i have again got this error during setup: https://pastebin.com/kDMJu07Q Do you wanna fix it too?

Definitely

dkayiwa commented:
Also, when i run the initial setup with jetty while having the legacyui module in the modules folder, on completing the installation, i get the `If you are seeing this page, it means that the OpenMRS Platform is running successfully, but no user interface module is installed. Learn from the available resources below.` page instead of the login page of the legacyui. I have to restart in order to get this page.

mherman22 commented:
> Also, when i run the initial setup with jetty while having the legacyui module in the modules folder, on completing the installation, i get the `If you are seeing this page, it means that the OpenMRS Platform is running successfully, but no user interface module is installed. Learn from the available resources below.` page instead of the login page of the legacyui. I have to restart OpenMRS in order to get this page.

That page should come on the very first run using jetty (where the .OpenMRS folder gets created).

After adding the legacyui omod to the modules folder and do a rerun of jetty, it should take you to the login page.

If that's not the behavior you're getting then that needs to be fixed. I will dig deep into that.

dkayiwa commented:
It does take me to the login page on a rerun. But it still does not take me there on at the completion of the setup during the first run.

mherman22 commented:
> It does take me to the login page on a rerun. But it still does not take me there at the completion of the setup during the first run.

I have done the following to try and reproduce this:
- i have deleted the .OpenMRS directory
- then did a repackage of core and then did either of mvn jetty:run or mvn cargo:run (i have tried them all)
- used the simple step and i landed on the page `If you are seeing this page, it means that the OpenMRS Platform is running successfully, but no user interface module is installed. Learn from the available resources below`
- I then downloaded the legacyui omod (1.22.0) and dropped it in the modules folder
- Did a rerun of either mvn jetty:run or mvn cargo:run
- I landed on the login page

maybe i am doing it differently, is this what you would expect? or how should i reproduce it?

dkayiwa commented:
When i run the setup wizard i already had the modules folder with the legacyui module.

mherman22 commented:
> * I then downloaded the legacyui omod (1.22.0) and dropped it in the modules folder



> When i run the setup wizard i already had the modules folder with the legacyui module.

Oh finally, i have reproduced it with mvn jetty:run. I wonder why, let me create a ticket a ticket for it and add it on my todos aswell. Thank you for trying this out.

dkayiwa commented:
100% reproducible. All you need is have the legacyui in the modules folder and then run the set up wizard with jetty.

dkayiwa commented:
@mherman22 i have consistently noticed that when doing an update and to the maintenance mode, after login and the changesets start executing, i get the maintenance mode login screen again, while the changesets are still executing. The behaviour that i expect is to keep seeing the changesets running until they are done. FWIW, i am running with mvn jetty:run

mherman22 commented:
> @mherman22 i have consistently noticed that when doing an update and to the maintenance mode, after login and the changesets start executing, i get the maintenance mode login screen again, while the changesets are still executing. The behaviour that i expect is to keep seeing the changesets running until they are done. FWIW, i am running with mvn jetty:run

I have been looking into since yesterday, i should have a PR soon if not today

mherman22 commented:
@dkayiwa i have added a commit at https://github.com/openmrs/openmrs-core/pull/5076 that attempts to keep the user authenticated throughout the entire process. I believe this was failing because initially an instance variable `authenticatedSuccessfully` was used to do this but it is lost between requests since each HTTP request gets a new instance of the filter, Instance variables don't persist between requests and the variable was reset to false on each new request.
